### ISSUE TITLE
Track low memory events

### ISSUE THREAD

**Issue by bidetofevil** [open-telemetry/opentelemetry-android#283]
Android's [ComponentCallback2](https://developer.android.com/reference/android/content/ComponentCallbacks2) interface allows for the app to listen for OS-level memory events that tell the app the process is trying to recover memory. Surfacing these warnings/alerts (as Events?) could be useful as it could be an indication the app is running low on resources so performance could be hampered.

**Comment by marandaneto** [comment]
It's worth adding a min. level configuration since not all levels are critical.
Also, worth adding a deboucer since this can be super noisy.
https://developer.android.com/topic/performance/memory

**Comment by breedx-splk** [comment]
I think this is a great idea. Contributions for this are very much welcome!

**Comment by aldoKelvianto** [comment]
I think this is a great idea as well! We can monitor when the app is running on low memory and identify the exact time when it happens.

However, I'm still not sure where to put this event in the context of OpenTelemetry Signals. We can put the low memory event as:

1. [Span Event](https://opentelemetry.io/docs/concepts/signals/traces/#span-events)

2. [Event in Log](https://opentelemetry.io/docs/concepts/signals/logs/#log-record)

I'm leaning towards using a Span Event, since we can get the full picture of the app's events by looking at Spans in the Trace.

What do you guys think?

**Comment by bidetofevil** [comment]
What signal is appropriate depends entirely on how sessions or being modelled in the app/library that is capturing the telemetry. For instance, in the Embrace SDK, this will be SpanEvent attached to the specific span we use to model a session, but this may not be appropriate for everyone.

The new `Events` signal seems to make the most sense to me in the long term - certainly a semconv that is attached to an Event makes sense.

In the context of this project, @breedx-splk probably has the best insight as to how best to model it, either now or in the future.

**Comment by breedx-splk** [comment]
Yeah, I agree with @bidetofevil that `Events` will be the way to model this. One way to tell is that the title of this issue even has the word "event" in it, and we think about these low memory points as events. 😄 

I believe the latest core java sdk release has the updated API that matches changes to the spec and semconv, and will allow rich objects in event payload bodies. Would be great to see what an implementation of these would look like, with respect to low memory.

**Comment by aldoKelvianto** [comment]
Cool! I just found out that `Events` support is already on development ([Event API Interface](https://opentelemetry.io/docs/specs/otel/logs/event-api/), [Semantic Convention for Events](https://opentelemetry.io/docs/specs/semconv/general/events/)), thanks 🙌

> One way to tell is that the title of this issue even has the word "event" in it, and we think about these low memory points as events. 😄

> The new Events signal seems to make the most sense to me in the long term - certainly a semconv that is attached to an Event makes sense.

Agree that `Events` is the most suitable signal for this kind of use case 👍

**Comment by Doohl** [comment]
I'd like to revive this issue with some thoughts:

Like mentioned above, we can use the [ComponentCallbacks2.onTrimMemory(int)](https://developer.android.com/reference/android/content/ComponentCallbacks2) method to get signals on memory events. For example:
1. `TRIM_MEMORY_RUNNING_MODERATE` -- the process is not an expendable background process, but the device is running moderately low on memory. 
2. `TRIM_MEMORY_RUNNING_LOW` -- the process is not an expendable background process, but the device is running low on memory.
3. `TRIM_MEMORY_RUNNING_CRITICAL` -- the process is not an expendable background process, but the device is running extremely low on memory and is about to not be able to keep any background processes running

Seems great. But! It has one glaring problem: seems like after API level 34, apps no longer get these granular memory notifications.

Instead, we'd have to use the [Memory Advice API](https://developer.android.com/games/sdk/memory-advice/overview), which is currently in Beta. The Memory Advice API gives us a few signals:
1. `MEMORYADVICE_STATE_OK`
2. `MEMORYADVICE_STATE_APPROACHING_LIMIT`
3. `MEMORYADVICE_STATE_CRITICAL`

I think we can join these two signals together and get at least two different types of Events out of it:
1. `memory.running_low` that maps to `TRIM_MEMORY_RUNNING_LOW` and `MEMORYADVICE_STATE_APPROACHING_LIMIT` (API 34+)
2. `memory.running_critical` that maps to `TRIM_MEMORY_RUNNING_CRITICAL` and `MEMORYADVICE_STATE_CRITICAL` (API 34+)

Open to thoughts!

**Comment by bidetofevil** [comment]
This seem interesting, though this [beta API](https://developer.android.com/games/sdk/memory-advice/overview) is only available on the `Android Games Development Kit`, and it's only supported for native apps (I guess it doesn't consider the JVM heap when firing these events).

Have you played with it to see how it functions?

**Comment by DavidGrath** [comment]
> Instead, we'd have to use the [Memory Advice API](https://developer.android.com/games/sdk/memory-advice/overview), which is currently in Beta

Deprecated 😔

**Comment by breedx-splk** [comment]
> > Instead, we'd have to use the [Memory Advice API](https://developer.android.com/games/sdk/memory-advice/overview), which is currently in Beta
> 
> Deprecated 😔

oh bummer! 😭 The [other recommended approaches](https://developer.android.com/games/optimize/memory-allocation#use-available-tools) aren't super compelling....

**Comment by bidetofevil** [comment]
To do this will require us to use heuristics like heap size paired with polling at the right moments to detect leaks, or leverage the `ProfilingManager` API and report the raw data async and have the analysis be done out of band on the server.

It's tricky and opinionated but could be really useful, but not straight forward to implement, design, and test, especially when we can't easily get production data. It is probably not a good candidate for this project at this point, but something that maybe a vendor like Embrace could look at, given they will have better means to validate.
